File: BombizeClient/Assets/Scripts/Network/UdpPacketDeliverer.cs
using System;
using System.IO;
using System.Collections.Generic;
using GameNetwork.Utilities;
using UnityEngine;

public class UdpPacketDeliverer
{
    private Action<MemoryStream> _onInputGameAction;
    public event Action<MemoryStream> OnInputGameAction
    {
        add => _onInputGameAction = value;
        remove
        {
            if(_onInputGameAction == value) _onInputGameAction = null;
        }
    }

    private UdpDriver _udpDriver;
    private uint _accessKey;
    private ushort _roomNumber;

    private uint _nextPackageNumber = 1;
    private uint _expectedPackageNumber = 1;
    private LinkedList<(uint number, byte[] rawData)> _unconfirmedPackets = new();
    private MemoryStream _currentOutStream = null;
    private MemoryStream _header = new();
    private struct Settings
    {
        public uint outConfirmationNumber;
        public bool needConfirmation;
        public Settings(uint outConfirmationNumber, bool needConfirmation)
        {
            this.outConfirmationNumber = outConfirmationNumber;
            this.needConfirmation = needConfirmation;
        }
    }
    private Settings _currentSettings = new(0, false);

    public UdpPacketDeliverer(UdpDriver udpDriver, uint accessKey, ushort roomNumber)
    {
        _udpDriver = udpDriver;
        _accessKey = accessKey;
        _roomNumber = roomNumber;

        _udpDriver.OnReceive += ProcessingInputPacket;
    }

    ~UdpPacketDeliverer() // temp
    {
        _udpDriver.OnReceive -= ProcessingInputPacket;
    }

    private void ProcessingInputPacket(byte[] rawData)
    {
        var inMStream = new MemoryStream(rawData, false);

        var inNumber = Utils.ReadUInt32(inMStream);
        var hasConfirmation = Utils.ReadByte(inMStream);
        if (hasConfirmation > 0)
        {
            var confirmationNumber = Utils.ReadUInt32(inMStream);
            ConfirmPacket(confirmationNumber);
        }

        //Debug.Log($"Input packet #{inNumber} ({hasConfirmation}). Expected: {_expectedPackageNumber}");

        if (inNumber == 0)
        {
            if (inMStream.Position < inMStream.Length)
                _onInputGameAction?.Invoke(inMStream);
        }
        else if (inNumber <= _expectedPackageNumber)
        {
            _currentOutStream = new();
            _currentSettings.outConfirmationNumber = inNumber;
            _currentSettings.needConfirmation = false;

            if (inNumber == _expectedPackageNumber && inMStream.Position < inMStream.Length)
            {
                _expectedPackageNumber++;
                _onInputGameAction?.Invoke(inMStream);
            }
        }
        // else throw away the input package
    }

    public MemoryStream GetCurrentOutStream()
    {
        return _currentOutStream ??= new();
    }

    public void SetConfirmOfCurrentPackage(bool confirm = true)
    {
        _currentSettings.needConfirmation |= confirm;
    }

    public void SendCurrentData()
    {
        if(_currentOutStream != null)
        {
            WriteInHeaderAccessData();

            if(_currentSettings.needConfirmation)
                Utils.WriteToStream(_header, _nextPackageNumber);
            else
                Utils.WriteToStream(_header, (uint)0);
            
            if(_currentSettings.outConfirmationNumber > 0)
            {
                Utils.WriteToStream(_header, (byte)1);
                Utils.WriteToStream(_header, _currentSettings.outConfirmationNumber);
            }
            else
            {
                Utils.WriteToStream(_header, (byte)0);
            }

            var sendData = JoinSendData(_header.ToArray(), _currentOutStream.ToArray());
            //var sendData = JoinSendData(_header.GetBuffer(), _currentOutStream.GetBuffer());
            _currentOutStream.Dispose();

            if(_currentSettings.needConfirmation)
            {
                _unconfirmedPackets.AddLast((_nextPackageNumber, sendData));
                _nextPackageNumber++;
            }
            _udpDriver.SendData(sendData);
            ResetSelfState();
        }
    }

    private byte[] JoinSendData(byte[] header, byte[] body)
    {
        byte[] result = new byte[header.Length + body.Length];
        Buffer.BlockCopy(header, 0, result, 0, header.Length);
        Buffer.BlockCopy(body, 0, result, header.Length, body.Length);
        return result;
    }

    private void ResetSelfState()
    {
        _currentOutStream = null;
        _header.SetLength(0);
        _currentSettings.needConfirmation = false;
        _currentSettings.outConfirmationNumber = 0;
    }

    private void WriteInHeaderAccessData()
    {
        Utils.WriteToStream(_header, _roomNumber);
        Utils.WriteToStream(_header, _accessKey);
    }

    public bool RepeatSend()
    {
        if(_unconfirmedPackets.Count == 0) return false;

        foreach (var item in _unconfirmedPackets)
        {
            _udpDriver.SendData(item.rawData);
        }
        return true;
    }

    private void ConfirmPacket(uint number)
    {
        var currentNode = _unconfirmedPackets.First;
        while (currentNode != null)
        {
            var nextNode = currentNode.Next;
            if (currentNode.Value.number == number)
            {
                _unconfirmedPackets.Remove(currentNode);
                break;
            }
            currentNode = nextNode;
        }
    }
}

File: BombizeServer/Bmbz/Network/UdpPacketDeliverer.cpp
#include "UdpPacketDeliverer.hpp"

namespace Bmbz
{
namespace Network
{
    void UdpPacketDeliverer::UpdateEndpoint(const boost::asio::ip::udp::endpoint& endpoint)
    {
        if(_endpoint != endpoint) _endpoint = endpoint;
    }

    bool UdpPacketDeliverer::IsConnection() const
    {
        return std::chrono::steady_clock::now() - _lastInputPacketTime <= _disconnectTime;
    }

    bool UdpPacketDeliverer::ProcessingInputPacket(boost::beast::flat_buffer& buffer)
    {
        _lastInputPacketTime = std::chrono::steady_clock::now();
        uint32_t inNumber;
        uint8_t hasConfirmation;
        uint32_t confirmationNumber{0};
        Utils::ReadFromBuffer(buffer, inNumber);
        Utils::ReadFromBuffer(buffer, hasConfirmation);
        
        if(hasConfirmation)
        {
            Utils::ReadFromBuffer(buffer, confirmationNumber);
            ConfirmPacket(confirmationNumber);
        }

        // std::cout << "Input UdpPacke #" << inNumber  << " | " << (int)hasConfirmation << " | " << confirmationNumber
        // << " --- " << _expectedPackageNumber << "\n";

        if(inNumber == 0)
        {
            return buffer.size() > 0;
        }
        else if(inNumber <= _expectedPackageNumber)
        {
            // prepare the response buffer
            _currentOutBuffer = _bufferPool.GetAndOccupyFreeBuffer();
            _currentSettings.outConfirmationNumber = inNumber;
            _currentSettings.needConfirmation = false;

            if(inNumber == _expectedPackageNumber)
            {
                _expectedPackageNumber++;
            }
            return buffer.size() > 0;
        }
        return false; // discard the packet and wait for the lost ones
    }

    void UdpPacketDeliverer::ConfirmPacket(uint32_t number)
    {
        for(auto it = _unconfirmedPackets.begin(); it != _unconfirmedPackets.end(); ++it)
            if(it->first == number)
            {
                _unconfirmedPackets.erase(it);
                break;
            }
    }

    boost::beast::flat_buffer& UdpPacketDeliverer::GetOutBuffer()
    {
        if(!_currentOutBuffer)
            _currentOutBuffer = _bufferPool.GetAndOccupyFreeBuffer();
        return _currentOutBuffer->body;
    }

    void UdpPacketDeliverer::SetConfirmOfCurrentPackage(bool confirm)
    {
        _currentSettings.needConfirmation |= confirm;
    }

    void UdpPacketDeliverer::SendCurrentData()
    {
        if(!_currentOutBuffer || _currentOutBuffer->body.size() == 0) return;

        if(_currentSettings.needConfirmation)
        {
            _currentOutBuffer->autoFree = false;
            Utils::WriteToBuffer(_currentOutBuffer->header, _nextPackageNumber);
            _unconfirmedPackets.emplace_back(_nextPackageNumber, _currentOutBuffer);
            _nextPackageNumber++;
        }
        else
        {
            Utils::WriteToBuffer(_currentOutBuffer->header, uint32_t{0});
        }
        if(_currentSettings.outConfirmationNumber > 0)
        {
            Utils::WriteToBuffer(_currentOutBuffer->header, uint8_t{1});
            Utils::WriteToBuffer(_currentOutBuffer->header, _currentSettings.outConfirmationNumber);
        }
        else
        {
            Utils::WriteToBuffer(_currentOutBuffer->header, uint8_t{0});
        }
        _udpMaster.SendData(_currentOutBuffer, _endpoint);
        ResetSelfState();
    }

    void UdpPacketDeliverer::ResetSelfState()
    {
        _currentOutBuffer.reset();
        _currentSettings.needConfirmation = false;
        _currentSettings.outConfirmationNumber = 0;
    }

    void UdpPacketDeliverer::RepeatSend()
    {
        if(_unconfirmedPackets.size() > 0)
        {
            for(auto& item : _unconfirmedPackets)
            {
                if(item.second->Sent())
                {
                    item.second->Sent(false);
                    _udpMaster.SendData(item.second, _endpoint);
                }
            }
        }
    }
}
}